File: 1162-as-far-from-land-as-possible/1162-as-far-from-land-as-possible.cpp
class Solution {
public:
    // bfs on each cell with value 1
    // When you can only go 4 directions, the (minimum) number of cells you travel from point A to point B is the Manhattan distance.



    int maxDistance(vector<vector<int>>& grid) {
        int r=grid.size(),c=grid[0].size(),steps=0;
        queue<pair<int,int>>q;
        
        for(int i =0;i<r;i++)
            for(int j=0;j<c;j++)
                if(grid[i][j]==1)
                    q.push({i-1,j}),q.push({i+1,j}), q.push({i,j-1}),q.push({i,j+1});
        while(!q.empty()){
            // in every iteration we are moving 1 step in bfs from every island
            int size = q.size();
            steps++;
            for(int k=0;k<size;k++){
                int i = q.front().first,j = q.front().second;
                q.pop();
                if(i>=0&&j>=0&&i<r&&j<c&&grid[i][j]==0){
                    grid[i][j]=steps;
                    q.push({i-1,j}),q.push({i+1,j}), q.push({i,j-1}),q.push({i,j+1});
                }
                    
            }
        }
        return steps==1?-1:steps-1; // if steps=0 then also return 0-1=0,  
    }
};

File: 1162-as-far-from-land-as-possible/1162-as-far-from-land-as-possible.py
from collections import deque
class Solution:
    def maxDistance(self, grid: List[List[int]]) -> int:
        rows,cols,steps=len(grid),len(grid[0]),0
        queue=collections.deque()
        for x in range(rows):
            for y in range(cols):
                if grid[x][y]==1:
                    queue.appendleft((x,y+1))
                    queue.appendleft((x,y-1))
                    queue.appendleft((x+1,y))
                    queue.appendleft((x-1,y))
        while queue:
            size=len(queue)
            steps+=1
            for k in range(size):
                x,y=queue[-1]
                queue.pop()
                if x>=0 and y>=0 and x<rows and y<cols and grid[x][y]==0:
                    grid[x][y]=steps
                    queue.appendleft((x,y+1))
                    queue.appendleft((x,y-1))
                    queue.appendleft((x+1,y))
                    queue.appendleft((x-1,y))
        # print("steps",steps)
        if steps==1:
            return -1 
        return steps-1
                                        
                    
                    
                
            
            